kafkaserver/handler.go

```diff
@@ -582,39 +579,66 @@ func (n *NewHandler) HandleSyncGroupRequest(_ *protocol.RequestHeader, req *prot
+func (n *connection) HandleApiVersionsRequest(_ *protocol.RequestHeader, _ *protocol.ApiVersionsRequest, completionFunc func(resp *protocol.ApiVersionsResponse) error) error {
 	var resp protocol.ApiVersionsResponse
 	resp.ApiKeys = protocol.SupportedAPIVersions
 	return completionFunc(&resp)
 }
 
-func (n *NewHandler) SaslHandshakeRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslHandshakeRequest) *protocol.SaslHandshakeResponse {
+func (n *connection) SaslHandshakeRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslHandshakeRequest) *protocol.SaslHandshakeResponse {
 	var resp protocol.SaslHandshakeResponse
 	resp.ErrorCode = errorCode
 	return &resp
 }
 
-func (n *NewHandler) HandleSaslHandshakeRequest(_ *protocol.RequestHeader, _ *protocol.SaslHandshakeRequest, completionFunc func(resp *protocol.SaslHandshakeResponse) error) error {
+func (n *connection) HandleSaslHandshakeRequest(_ *protocol.RequestHeader, req *protocol.SaslHandshakeRequest, completionFunc func(resp *protocol.SaslHandshakeResponse) error) error {
 	var resp protocol.SaslHandshakeResponse
-	// For now, we don't implement any mechanisms, this is just a placeholder
+	conversation, ok, err := n.s.saslAuthManager.CreateConversation(*req.Mechanism)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp.ErrorCode = protocol.ErrorCodeUnsupportedSaslMechanism
+	} else {
+		n.saslConversation = conversation
+	}
+	saslScram256 := auth.AuthenticationSaslScramSha256
+	saslScram512 := auth.AuthenticationSaslScramSha512
+	resp.Mechanisms = []*string{&saslScram256, &saslScram512}
 	return completionFunc(&resp)
 }
 
-func (n *NewHandler) SaslAuthenticateRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslAuthenticateRequest) *protocol.SaslAuthenticateResponse {
+func (n *connection) SaslAuthenticateRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslAuthenticateRequest) *protocol.SaslAuthenticateResponse {
 	var resp protocol.SaslAuthenticateResponse
 	resp.ErrorCode = errorCode
 	return &resp
 }
 
-func (n *NewHandler) HandleSaslAuthenticateRequest(_ *protocol.RequestHeader, req *protocol.SaslAuthenticateRequest, completionFunc func(resp *protocol.SaslAuthenticateResponse) error) error {
+func (n *connection) HandleSaslAuthenticateRequest(_ *protocol.RequestHeader, req *protocol.SaslAuthenticateRequest, completionFunc func(resp *protocol.SaslAuthenticateResponse) error) error {
 	var resp protocol.SaslAuthenticateResponse
-	// Always fail, just a placeholder
-	resp.ErrorCode = protocol.ErrorCodeSaslAuthenticationFailed
+	conv := n.saslConversation
+	if conv == nil {
+		resp.ErrorCode = protocol.ErrorCodeIllegalSaslState
+		msg := "SaslAuthenticateRequest without a preceding SaslAuthenticateRequest"
+		resp.ErrorMessage = &msg
+	} else {
+		saslRespBytes, complete, failed := conv.Process(req.AuthBytes)
+		if failed {
+			resp.ErrorCode = protocol.ErrorCodeSaslAuthenticationFailed
```

Comment:
NIT: I see the response has an ErrorMessage, should we create a follow-up issue maybe to populate the message, specially if Kafka does that as well?

Reply:
For auth failed, what would a useful error message be? If user is not found, we can't say user not found, as can tell a malicious user who is trying users, that the user is valid or not. If password not correct, then not much we can say either.